### ISSUE TITLE
Sik-Ho Tsang | Review: DCGAN -- Deep Convolutional Generative Adversarial Network (GAN).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#58]
Sik-Ho Tsang. [Review: DCGAN — Deep Convolutional Generative Adversarial Network (GAN)](https://sh-tsang.medium.com/review-dcgan-deep-convolutional-generative-adversarial-network-gan-ec390cded63c).

**Comment by NorbertZheng** [comment]
## Overview
Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks.

In this story, Deep Convolutional Generative Adversarial Network (DCGAN), by Indico Research and Facebook AI Research (FAIR), is reviewed. With DCGAN, **a hierarchy of representations** is learned from object parts to scenes in both the generator and discriminator. This is a paper in 2016 ICLR with about 6000 citations.

**Comment by NorbertZheng** [comment]
## A Set of Constraints for Stable Training
### All convolutional net replaces deterministic spatial pooling functions (such as max pooling) with strided convolutions (i.e. dilation convolution).
 - The generator learns its own spatial downsampling (relative to the dimension of input space) itself using convolution.
 - Similarly, the discriminator learns its own spatial upsampling (relative to the dimension of embedding space).

### Eliminating Fully Connected Layers.
 - The first layer of the generator, which takes a uniform noise distribution $Z$ as input, could be called fully connected as it is just a matrix multiplication, but the result is reshaped into a 4-dimensional tensor and used as the start of the convolution stack.
 - For the discriminator, the last convolution layer is flattened and then fed into a single sigmoid output.
 - But there are no fully connected layers for all hidden layers.
Bye bye fully connected layers, haha~

### [Batch Normalization (BN)](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651).
 - [BN](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) stabilizes learning by normalizing the input to each unit to have zero mean and unit variance..
 - However, ~~directly applying [BN](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) to all layers~~ resulted in sample oscillation and model instability.
 - This was avoided by not applying [BN](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) to the generator output layer and the discriminator input layer.

### Activation Functions.
 - The ReLU activation is used in the generator with the exception of the output layer which uses the Tanh function.
 - Within the discriminator, it is found that the leaky rectified activation (LeakyReLU) works well.

**Comment by NorbertZheng** [comment]
## Network Architecture
### Generator
![image](https://user-images.githubusercontent.com/37235023/216622142-833e716e-6963-46ab-9a21-183c2b513f09.png)
 - The generator is as shown above. Only convolution, no fully connected layers according to the constraints.
### Discriminator
![image](https://user-images.githubusercontent.com/37235023/216622318-93cc8bb5-a0ca-4817-ae1b-357163b43028.png)
 - The discriminator is just like a inverse of generator.

**Comment by NorbertZheng** [comment]
## Experimental Results
 - Three datasets are trained: LSUN, ImageNet, and a newly assembled faces dataset.
### LSUN
![image](https://user-images.githubusercontent.com/37235023/216622509-51152e90-3680-4928-94c6-86e882997d74.png)
Generated bedrooms after one epoch.
![image](https://user-images.githubusercontent.com/37235023/216622595-e01dfdbd-a7e4-4ba5-9919-8d1da0d7a00d.png)
Generated bedrooms after five epochs of training.

 - A model is trained on the LSUN bedrooms dataset containing a little over 3 million training examples.
 - The model is not producing high quality samples via simply overfitting/memorizing training examples.

**Comment by NorbertZheng** [comment]
### DCGAN Trained on ImageNet, Tested on CIFAR10 & SVHN
![image](https://user-images.githubusercontent.com/37235023/216622832-36eb4784-2f50-4a3b-96d5-2c4498e8f7c4.png)
Accuracy (%) on CIFAR10.

 - DCGAN is trained on ImageNet-1k and then
    - use the discriminator’s convolutional features from all layers, maxpooling each layers representation to produce a 4×4 spatial grid.

   These features are then flattened and concatenated to form a 28672 dimensional vector and a regularized linear L2-SVM classifier is trained on top of them.
 - This achieves 82.8% accuracy, outperforms all K-means based approaches. Error Rate (%) on SVHN is as follows:
   ![image](https://user-images.githubusercontent.com/37235023/216623250-77f2fd38-e29d-404d-ad3b-20f446d9c838.png)
 - A purely supervised CNN is trained but only 28.87% error rate is obtained.
 - But using DCGAN, the same CNN, 22.8% error rate is obtained.

**Comment by NorbertZheng** [comment]
### Walking in the Latent Space
![image](https://user-images.githubusercontent.com/37235023/216623506-0da7eac7-f7ee-4186-8732-de6cce2c7a2c.png)
Interpolation between a series of 9 random points in Z show that the space learned has smooth transitions.
 - The vector $Z$ actually is a $n$ dimensional vector in a $n$ dimensional space.
 - If interpolation is performed between 2 $z$ vectors, a gradual change can be seen as shown above.
 - Walking in this latent space results in **semantic changes** to the image generations (such as objects being added and removed), we can reason that the model has learned relevant and interesting representations.

**Comment by NorbertZheng** [comment]
### Visualizing the Discriminator Features
![image](https://user-images.githubusercontent.com/37235023/216623856-3d0e4f44-7904-4d5d-85a0-cc35726b8ad3.png)
Left: Random Filter Baseline, Right: Trained Filters.
 - The above figure shows that the features learnt by the discriminator activate on typical parts of a bedroom, like beds and windows.

**Comment by NorbertZheng** [comment]
### Forgetting to Draw Certain Objects
![image](https://user-images.githubusercontent.com/37235023/216624036-ec37ac65-8506-4ffa-8ece-0892d1c31fa6.png)
First Row: Models Without Dropping “Window” Filters, Second Row: Models With Dropping “Window” Filters.
 - By dropping “Window” filter, some windows are removed, others are transformed into objects with similar visual appearance such as doors and mirrors.
 - Although visual quality decreased, overall scene composition stayed similar, suggesting the generator has done a good job disentangling scene representation from object representation.

**Comment by NorbertZheng** [comment]
### Vector Arithmetic on Face Samples
![image](https://user-images.githubusercontent.com/37235023/216624280-f10dfdbf-4445-4b9b-9bd2-ce7d3083bbd0.png)
Smiling Woman — Neutral Woman + Neutral Man = Smiling Man.

![image](https://user-images.githubusercontent.com/37235023/216624349-482145d8-bb83-4a95-b646-1d132bb96225.png)
Man With Glasses — Man Without Glasses + Woman Without Glasses = Woman With Glasses.

 - Simple arithmetic operations revealed rich linear structure in representation space.
 - e.g.: vector(”King”) — vector(”Man”) + vector(”Woman”) can result in a vector whose nearest neighbor was the vector for Queen.
 - Experiments working on only single samples per concept were unstable, but **averaging the $Z% vector for three exemplars showed consistent and stable generations that semantically obeyed the arithmetic**.
 - As shown in the figures above, for each column, the $Z$ vectors of samples are averaged. Arithmetic was then performed on the mean vectors creating a new vector $Y$. This $Y$ is fed into the generator as input.
 - A uniform noise sampled with scale ±0.25 was added to $Y$ to produce the other 8 samples.

![image](https://user-images.githubusercontent.com/37235023/216624775-151d33c5-cd86-41dd-8b96-1528c53cc1ab.png)
A ”turn” vector was created from four averaged samples of faces looking left vs looking right.
 - Face pose can also be modeled linearly in $Z$ space.

![image](https://user-images.githubusercontent.com/37235023/216624913-c84cc531-639e-4889-9103-429cc668407d.png)
Vector Arithmetic on Input Space.
 - In contrast, vector arithmetic on input space obtains poor results.

**Comment by NorbertZheng** [comment]
## Reference
[2016 ICLR] [DCGAN] [Unsupervised Representation Learning with Deep Convolutional Generative Adversarial Networks](https://arxiv.org/abs/1511.06434).

**Comment by NorbertZheng** [comment]
## Generative Adversarial Network
[[GAN](https://medium.com/@sh.tsang/review-gan-generative-adversarial-nets-gan-e12793e1fb75)] [[CGAN](https://medium.com/@sh.tsang/review-cgan-conditional-gan-gan-78dd42eee41)] [[LAPGAN](https://medium.com/@sh.tsang/review-lapgan-laplacian-generative-adversarial-network-gan-e87200bbd827)] [[DCGAN](https://medium.com/@sh.tsang/review-dcgan-deep-convolutional-generative-adversarial-network-gan-ec390cded63c)]
